### ISSUE TITLE
Blaze conditional expression?

### ISSUE THREAD

**Issue by kwmsmith** [blaze/blaze#1574]
This came up when a user wanted to compute a new column to segment integers into a handful of categories, computing on a redshift backend.  There's `.map()`, but that won't work for this backend.  I couldn't come up with a straightforward way to do it, and I kept wishing I had something like a `cond()` expression.

Example:

```python
In [36]: t = bz.symbol('t', 'var * {income: int32}')

In [37]: bz.transform(t,
    ...:              p = bz.cond((t.income < 55000), "A",
    ...:                          (t.income < 135000), "B",
    ...:                          (t.income < 187000), "C",
    ...:                          "else", "D"))
```

Thoughts?

**Comment by llllllllll** [comment]
numpy calls this `where` or maybe `choose`

```python
In [10]: np.where(np.array([True, False, True, False]), np.arange(4), -np.arange(4))
Out[10]: array([ 0, -1,  2, -3])

In [19]: np.choose((arr > 2).astype(int) + (arr >  4).astype(int), ('A', 'B', 'C'))
Out[19]: 
array(['A', 'A', 'B', 'B', 'C', 'C'], 
      dtype='|S1')
```

I do like the `case` style function you proposed since we can evaluate the bool exprs in order. How do you feel about something like:

```python
bz.case[
    t.income < 55000: 'A',
    t.income < 135000: 'B',
    t.income < 187000: 'C',
    True: 'D',
]
```
This makes the key:value relationship very clear imho.

In any of the APIs we choose I think `True` or an alias `otherwise = True` is preferable to `"else"` because it makes all the conditions boolean type and makes it easier to explain as:

> execute all the booleans in order and select the first alternative for which the conditon evaluates to `True`.

**Comment by postelrich** [comment]
I would vote something for the below where each case is a tuple, with a keyword argument for the default/else case.  `else` may be more pythonic as that is what is used for `if...elif...else`, the only way to do cond/case/switch in python.

```
bz.cond((t.income < 55000, "A"),
              (t.income < 135000, "B"),
              (t.income < 187000, "C"),
              else="D")
```

Also how would this be used outside of a transform?  What would it return outside of transform, a single column?

Other names could be `bz.bin` or `bz.bucket`.

**Comment by necaris** [comment]
I'm -1 on the dictionary-style syntax, since (to me at least) that implies a dictionary implementation, which loses ordering, but I do like the `case` or `where` keywords as an alternative to `cond`.

Perhaps `where` is best, since it's familiar from both NumPy and SQL-style usages, while `cond` is more Lisp-specific?

**Comment by kwmsmith** [comment]
Yes, I thought of `np.where()` and `np.choose()` -- my preference at this point would be to make sure `cond` or `case` handles the two-branch scenario well and not clutter the namespace, but if there's a strong case to be made for a separate `where()` (or other name), then I'm all ears.

> How do you feel about something like: ...

Hmm, I don't know -- it doesn't strike me as getting us that much more syntax-wise than the original version above.  It's an interesting way to massage `__getitem__` and slices, and it reminds me of the `datashape.R['a': 'int32', 'b': 'string']` syntactic sugar to create Record DSes, as well as the `numpy.r_`, `numpy.c_` and other convenience constructors.  I still don't know if I love it or not.

I'm not wild about my original `"else"` as the catch-all predicate; I see an argument for `True`, and / or an `otherwise` alias.

What about requiring strict `bool` type for the predicates vs. being more flexible and supporting truthy things?

Lastly, what about generalizing the value slots to support expressions there as well?  Would we need to ensure that the value expression is only evaluated on the rows for which the predicate is True?

**Comment by kwmsmith** [comment]
> Lastly, what about generalizing the value slots to support expressions there as well? Would we need to ensure that the value expression is only evaluated on the rows for which the predicate is True?

To answer my own question -- from a correctness point of view, since the value expression is an expression and side-effect-free, then we can leave this as a future optimization, and whether the expr is evaluated on all rows or only the rows for which predicate is true could be backend specific.

**Comment by kwmsmith** [comment]
@postelrich unfortunately `else` is a reserved keyword, part of Python's syntax, so we can't use it as a kwarg:

```python
In [23]: def foo(a, b, else=10):
  File "<ipython-input-23-4ee946da1eaf>", line 1
    def foo(a, b, else=10):
                     ^
SyntaxError: invalid syntax
```

```python
In [24]: def foo(*args, **kwargs):
    ...:     print(args, kwargs)
    ...:

In [25]: foo(10, 20, else=30)
  File "<ipython-input-25-df88f1c9417e>", line 1
    foo(10, 20, else=30)
                   ^
SyntaxError: invalid syntax
```

**Comment by necaris** [comment]
@kwmsmith How about `otherwise`, as @llllllllll suggests -- or `default` which is pretty common in Python APIs I've seen.

**Comment by kwmsmith** [comment]
+1 for `otherwise` as a simple alias for `True`.  We can document just `otherwise`, kind of like `np.newaxis` is an alias for `None`.

**Comment by postelrich** [comment]
@kwmsmith good point, I would vote for default.
